### PR TITLE
COSE envelope implementation updates

### PR DESCRIPTION
**Pull request by shizhMSFT**
Tracking the merger of the [cose](https://github.com/notaryproject/notation-core-go/tree/cose) branch into `main`.

### CODE REVIEW

File: signature/signer.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	PrivateKey() crypto.PrivateKey
+}
+
+type signer struct {
+	keySpec KeySpec
+	key     crypto.PrivateKey
+	certs   []*x509.Certificate
+}
+
+func NewSigner(certs []*x509.Certificate, key crypto.PrivateKey) (Signer, error) {
```
Comment (shizhMSFT) [inline signature/signer.go:27]: @binbin-li Should we make it
```suggestion
func NewLocalSigner(certs []*x509.Certificate, key crypto.PrivateKey) (LocalSigner, error) {
```

/ccc @chloeyin

Comment (binbin-li) [reply signature/signer.go:27]: that does make more sense since we only need to create local signers.

**Comment by JeyJeyGao** [comment]
I will work on this task.

File: signature/algorithm.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		return crypto.SHA512
+	}
+	return 0
+}
+
+// ExtractKeySpec extracts keySpec from the signing certificate
+func ExtractKeySpec(signingCert *x509.Certificate) (KeySpec, error) {
+	switch key := signingCert.PublicKey.(type) {
+	case *rsa.PublicKey:
+		switch bitSize := key.Size() << 3; bitSize {
```
Comment (qmuntal) [inline signature/algorithm.go:56]: Why `<<3`? SHould be necessary.

**Comment by SteveLasker** [comment]
@binbin-li, @shizhMSFT, we have a set of DCO errors. Can we get these fixed up

**Comment by shizhMSFT** [comment]
> @binbin-li, @shizhMSFT, we have a set of DCO errors. Can we get these fixed up?

@SteveLasker This PR is just show the working progress of COSE, and will be split to multiple PRs to be reviewed and merged. Once those PRs are merged, this PR will be closed.

File: signature/algorithm.go (hunk 2 above)
Comment (shizhMSFT) [reply signature/algorithm.go:56]: [Size()](https://pkg.go.dev/crypto/rsa#PublicKey.Size) returns in bytes but we want in bits.

File: signature/errors.go [hunk 3]
```diff
@@ -107,3 +107,13 @@ type UnsupportedSignatureFormatError struct {
...
 }
+
+// EnvelopeKeyRepeatedError is used when repeated key name found in the envelope.
+type EnvelopeKeyRepeatedError struct {
+	Key string
+}
+
+// Error returns the formatted error message
+func (e *EnvelopeKeyRepeatedError) Error() string {
+	return fmt.Sprintf("repeated key: `%s` exists in the envelope.", e.Key)
```
Comment (qmuntal) [inline signature/errors.go:118]: The %q verb automagically quotes the string. Use it instead of \`%s\`.

Comment (shizhMSFT) [reply signature/errors.go:118]: @binbin-li Could you address this?

Comment (binbin-li) [reply signature/errors.go:118]: Sure, I'll do it.

**Comment by shizhMSFT** [comment]
This PR is succeeded by #72 and #75

**Pull request closed without merging**